File: tasks/dfsConnectedComponents.cpp
int dfsConnectedComponents(std::vector<std::vector<bool> > matrix) {

  struct Helper {
    std::vector<bool> visited;
    std::vector<std::vector<bool> > matrix;

    void dfs(int currentVertex) {
      visited[currentVertex] = true;
      for (int nextVertex = 0; nextVertex < matrix[0].size(); nextVertex++) {
        if (!matrix[currentVertex][nextVertex] || visited[nextVertex])
          continue;
        dfs(nextVertex);
      }
    }
  };

  std::vector<bool> visited(matrix.size(), false);

  Helper h;
  h.visited = visited;
  h.matrix = matrix;

  int componentsCount = 0;

  for (int startVertex = 0; startVertex < matrix.size(); startVertex++) {
    if (!h.visited[startVertex]) {
      h.dfs(startVertex);
      componentsCount++;
    }
  }

  return componentsCount;
}

File: tasks/dfsConnectedComponents.js
function dfsConnectedComponents(matrix) {

  var dfs = function(currentVertex, visited) {
    visited[currentVertex] = true;
    for (var nextVertex = 0; nextVertex < matrix[0].length; nextVertex++) {
      if (!matrix[currentVertex][nextVertex] || visited[nextVertex])
        continue;
      dfs(nextVertex, visited);
    }
  }

  var visited = [];
  var componentsCount = 0;

  for (var i = 0; i < matrix.length; i++) {
    visited.push(false);
  }

  for (var startVertex = 0; startVertex < matrix.length; startVertex++) {
    if (!visited[startVertex]) {
      dfs(startVertex, visited);
      componentsCount++;
    }
  }

  return componentsCount;
}
